src/aws/sns-calls.js

```diff
@@ -17,6 +17,103 @@
+        .then(topicAttributesResponse => {
+            console.log(topicAttributesResponse);
+            let policy = topicAttributesResponse.Attributes.Policy;
+            if (policy) {
+                let policyDoc = JSON.parse(policy);
+                return policyDoc;
+            }
+            else {
+                return null; //No policy defined
+            }
+        });
+}
+
+exports.addSnsPermission = function (topicArn, sourceArn, policyStatement) {
+    let sns = new AWS.SNS({ apiVersion: '2010-03-31' });
+
+    return getSnsPolicyDoc(sns, topicArn)
+        .then(policyDoc => {
+            if (policyDoc) { //Update existing policy
+                policyDoc.Statement.push(policyStatement);
+            }
+            else { //Create new policy
+                policyDoc = {
+                    "Version": "2012-10-17",
+                    "Statement": [policyStatement]
+                }
+            }
+
+            var setTopicAttributesParams = {
+                AttributeName: 'Policy',
+                TopicArn: topicArn,
+                AttributeValue: JSON.stringify(policyDoc)
+            };
+            winston.verbose(`Adding permission to SNS topic ${topicArn} from ${sourceArn}`);
+            return sns.setTopicAttributes(setTopicAttributesParams).promise()
+                .then(topicAttributesResponse => {
+                    winston.verbose(`Added permission to SNS Topic ${topicArn} from ${sourceArn}`);
+                    return topicAttributesResponse;
+                });
+        });
+}
+
+exports.getSnsPermission = function (topicArn, policyStatementToGet) {
+    let sns = new AWS.SNS({ apiVersion: '2010-03-31' });
+    winston.verbose(`Attempting to find permission in policy doc for ${topicArn}`);
+    return getSnsPolicyDoc(sns, topicArn)
+        .then(policyDoc => {
+            if (policyDoc) {
+                let permissionFromPolicyDoc = getPermissionInPolicyDoc(policyDoc, policyStatementToGet);
+                if (permissionFromPolicyDoc) {
+                    winston.verbose(`Found permission in policy doc for ${topicArn}`);
+                    return permissionFromPolicyDoc;
+                }
+                else {
+                    winston.verbose(`Permission does not exist in policy doc for ${topicArn}`);
+                    return null;
+                }
+            }
+            else {
+                winston.verbose(`No policy defined for ${topicArn}`);
+                return null;
+            }
+        });
+}
+
+exports.addSnsPermissionIfNotExists = function (topicArn, sourceArn, policyStatement) {
+    return exports.getSnsPermission(topicArn, policyStatement)
+        .then(permission => {
+            if (!permission) {
+                return exports.addSnsPermission(topicArn, sourceArn, policyStatement)
+                    .then(() => {
+                        return exports.getSnsPermission(topicArn, policyStatement);
```

Comment:
Is the thing returned from addSnsPermission the same as what's returned from getSnsPermission? Often AWS will return the same object from a create that they would from a get, so if that's the case here then we can remove this call and just return what comes back from the create.

I have seen cases where AWS didn't do this however, forcing me to call get after a create like you're doing here.

Reply:
I believe they are different as there is no getSnsPermission. This function is a wrapper. It calls getTopicAttributes and then parses that for the PolicyDoc. This could be confusing too as there is a call with sns called addPermission it  is used for adding permissions across accounts. In this case we actually just call setTopicAttributes to add a policy doc. and getTopicAttributes to retrieve the policy doc.

Reply:
Oh yeah that rings a bell, sorry I had forgotten how the permission in SNS are part of the topic attributes.